### PR TITLE
Create FAB provider and move FAB auth manager in it

### PR DESCRIPTION
**Pull request by vincbeck**
As a result of the [lazy consensus](https://lists.apache.org/thread/o457svvnsn4xdnb7j79jyqsxvldj6qmz), this PR creates a new provider: fab. This new provider contains (and should always only contain?) the FAB manager that is in core Airflow for now. 

It basically moves files from `airflow.auth.managers.fab` to `airflow.providers.fab.auth_manager`.

What this PR does NOT include and will be done in following PRs:
- Pre-install the FAB provider in Airflow
- Remove `flask-login` as dependency in core Airflow (only some tests in core Airflow are using it now, should be pretty easy to remove)
- Move FAB auth manager related configuration from core Airflow config to provider config
- Move FAB auth manager related documentation from core Airflow documentation to provider documentation
- Create documentation related to FAB auth manager

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -525,6 +524,7 @@ repos:
           ^airflow/providers/apache/spark/hooks/|
           ^airflow/providers/apache/spark/operators/|
           ^airflow/providers/exasol/hooks/exasol.py$|
+          ^airflow/providers/fab/auth_manager/security_manager/|
```
Comment (ferruzzi) [inline .pre-commit-config.yaml:534]: [Non-blocking]    Interesting.  I wonder why so many are exempt, we should maybe work on getting this list trimmed down in another PR.

File: setup.cfg [hunk 2]
```diff
@@ -103,6 +103,7 @@ install_requires =
     # NOTE! When you change the value here, you also have to update flask-appbuilder[oauth] in setup.py
     flask-appbuilder==4.3.9
     flask-caching>=1.5.0
+    # We should be able to remove flask-login. It is still used in core Airflow but in tests. Looks feasible.
```
Comment (ferruzzi) [inline setup.cfg:106]: It's not entirely removed, just moved into the provider, right?

File: docs/apache-airflow-providers-fab/index.rst [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+`Flask App Builder <https://flask-appbuilder.readthedocs.io/>`__
+
+
+Release: 1.0.0
+
+Provider package
+----------------
+
+This is a provider package for optional Airflow components using ``Flask App Builder``.
+All classes for this provider package are in ``airflow.providers.fab`` python module.
```
Comment (ferruzzi) [inline docs/apache-airflow-providers-fab/index.rst]:
```suggestion
All classes for this provider package are in the ``airflow.providers.fab`` python module.
```

File: docs/apache-airflow-providers-fab/index.rst [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+Installation
+------------
+
+You can install this package on top of an existing Airflow 2 installation (see ``Requirements`` below)
+for the minimum Airflow version supported) via ``pip install apache-airflow-providers-fab``
```
Comment (ferruzzi) [inline docs/apache-airflow-providers-fab/index.rst]: Not sure what happened here, I think this was your intent

```suggestion
You can install this package on top of an existing Airflow 2 installation (see ``Requirements`` below
for the minimum Airflow version supported) via ``pip install apache-airflow-providers-fab``
```

File: airflow/providers/fab/auth_manager/security_manager/override.py [hunk 5]
```diff
@@ -0,0 +1,2668 @@
+#
```
Comment (ferruzzi) [inline airflow/providers/fab/auth_manager/security_manager/override.py:1]: I ran this file and the old `airflow/auth/managers/fab/security_manager/override.py` though a diff, for easier review:  https://editor.mergely.com/kNuiVsA7

Changes are mainly to import paths and comments.

File: airflow/providers/fab/auth_manager/security_manager/override.py [hunk 6]
```diff
@@ -0,0 +1,2668 @@
...
+    userdbmodelview = CustomUserDBModelView
+    resetmypasswordview = CustomResetMyPasswordView
+    resetpasswordview = CustomResetPasswordView
+    userinfoeditview = CustomUserInfoEditView
+    userldapmodelview = CustomUserLDAPModelView
+    useroauthmodelview = CustomUserOAuthModelView
+    userremoteusermodelview = CustomUserRemoteUserModelView
+    useroidmodelview = CustomUserOIDModelView
+    userstatschartview = CustomUserStatsChartView
+
```
Comment (ferruzzi) [inline airflow/providers/fab/auth_manager/security_manager/override.py:182]: In the following block: I see that you just used it how it was in the old file (with some corrections), but I feel like the comment should be above the assignment it is describing.  I'll leave it up to you though, I don't feel too strongly about it.

**Review by ferruzzi — APPROVED** [review]
Left a few notes, but nothing blocking.   Looks good.

File: setup.cfg (hunk 2 above)
Comment (vincbeck) [reply setup.cfg:106]: Correct

File: .pre-commit-config.yaml (hunk 1 above)
Comment (vincbeck) [reply .pre-commit-config.yaml:534]: The issue is, very often it leads to breaking change (the method/function name needs to be changed). But still, you're right we could rename them and deprecate the others

File: airflow/providers/fab/auth_manager/security_manager/override.py [hunk 7]
```diff
@@ -2150,7 +2150,7 @@ def get_oauth_user_info(self, provider: str, resp: dict[str, Any]) -> dict[str,
             log.debug("User info from Azure: %s", me)
             # https://learn.microsoft.com/en-us/azure/active-directory/develop/id-token-claims-reference#payload-claims
             return {
-                "email": me["email"],
+                "email": me["upn"] if "upn" in me else me["email"],
```
Comment (ferruzzi) [inline airflow/providers/fab/auth_manager/security_manager/override.py]:
```suggestion
                "email": me.get("upn", me["email"]),
```

```
In [1]: me = {"email": 2}

In [2]: me.get("upn", me["email"])
Out[2]: 2

In [3]: me = {"upn": 1, "email": 2}

In [4]: me.get("upn", me["email"])
Out[4]: 1

```

**Comment by eladkal** [comment]
I may be late to the party but...
is `fab` the right name? 
Shouldn't it be the official name `FlaskAppBuilder` ?

**Comment by vincbeck** [comment]
CI is green :D

**Comment by potiuk** [comment]
🤞

**Comment by potiuk** [comment]
Wooohoooo. Compat tests passed :)

**Comment by potiuk** [comment]
Nope. It was https://github.com/apache/airflow/pull/36168 - rebase and it should be gone (and my PR also added different behaviour when constraint generation will fail - it will fail at constraint generation time instead of when they should be used.

**Comment by vincbeck** [comment]
CLI is green! (thanks @potiuk for the help!)

**Comment by potiuk** [comment]
Woohoooo!

**Comment by ephraimbuddy** [comment]
I think this move should have been part of 2.8.0. Another option would be to release it as part of 2.8.1 because if there's any bug on FAB that needs fixing in 2.8, we won't be able to do it because of this move.

**Comment by potiuk** [comment]
> I think this move should have been part of 2.8.0. Another option would be to release it as part of 2.8.1 because if there's any bug on FAB that needs fixing in 2.8, we won't be able to do it because of this move.

Well, yes and no. I think it will be very easy to apply cherry-picks from main in those cases as this was generally just moving the code - because we do not touch this code unless there is migration to newer FAB.

**Comment by potiuk** [comment]
I am happy to do those cherry-picks, if it makes things easier :)

**Comment by eladkal** [comment]
> I am happy to do those cherry-picks, if it makes things easier :)

Can this PR be cherry picked to 2.8 branch?
It changed import path in existed migration script `airflow/migrations/versions/0073_2_0_0_prefix_dag_permissions.py`
to import from fab provider but fab provider can be used only with airflow 2.9.0
If we cherry pick this PR won't it fail any airflow db upgrade ?

**Comment by ephraimbuddy** [comment]
> > I am happy to do those cherry-picks, if it makes things easier :)
> 
> Can this PR be cherry picked to 2.8 branch? It changed import path in existed migration script `airflow/migrations/versions/0073_2_0_0_prefix_dag_permissions.py` to import from fab provider but fab provider can be used only with airflow>=2.9.0 If we cherry pick this PR to 2.8 branch thus it will be part of 2.8.x release, won't it fail any airflow db upgrade ?

We shouldn't add more PRs to 2.8 branch. It's late now. 
And concerning future cherrypicks, it's not to cherry-pick this particular one but I'm just thinking that if there's an issue in Auth manager in 2.8.0, how do we fix it for 2.8.1 since the authmanager is now moved to the provider. That's my worry but hopefully, we won't have issues.

**Comment by potiuk** [comment]
We can also revert this one and keep a PR open and merge it when we get closer to 2.9.0 - that's also an option @vincbeck ?

**Comment by ashb** [comment]
FYi @vincbeck @ferruzzi This broke the `AIRFLOW__WEBSERVER__UPDATE_FAB_PERMS` config setting (the old setting simply got ignored)

It should have been added to https://github.com/apache/airflow/blob/main/airflow/configuration.py#L325-L394 to automatically upgrade the config in place and issue a warning to users.

**Comment by vincbeck** [comment]
I guess the PR you mention is #36232. I was not aware of `deprecated_options`, pretty handy, thanks for the info. Though, when we lookup for `AIRFLOW__FAB__UPDATE_FAB_PERMS`, we also lookup for `AIRFLOW__WEBSERVER__UPDATE_FAB_PERMS` as a fallback. Thus, I am happy to add this option to `deprecated_options` but if something is broken today, I dont think it will resolve it

**Comment by vincbeck** [comment]
#40325

**Comment by vincbeck** [comment]
Oops, #40317 was already addressing the problem

**Pull request merged**